pkg/cache/gvk_caches.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+type GVKCachesOptions struct {
+	ObjectCacheFuncs map[client.Object]NewCacheFunc
+	DefaultCacheFunc NewCacheFunc
+}
+
+// GVKCachesBuilder - Builder function to create a new composite cache which contains
+// one or more caches based on a mapping of GVKs. Each GVK can be configured to
+// use a separate cache.
+func GVKCachesBuilder(gvkCachesOpts GVKCachesOptions) NewCacheFunc {
+	return func(config *rest.Config, opts Options) (Cache, error) {
+		opts, err := defaultOpts(config, opts)
+		if err != nil {
+			return nil, err
+		}
+
+		if gvkCachesOpts.DefaultCacheFunc == nil {
+			gvkCachesOpts.DefaultCacheFunc = New
+		}
+		defaultCache, err := gvkCachesOpts.DefaultCacheFunc(config, opts)
+		if err != nil {
+			return nil, err
+		}
+
+		gvkToCache := map[schema.GroupVersionKind]Cache{}
+		for obj, newCacheFunc := range gvkCachesOpts.ObjectCacheFuncs {
+			gvk, err := apiutil.GVKForObject(obj, opts.Scheme)
+			if err != nil {
+				return nil, err
+			}
+			gvkToCache[gvk], err = newCacheFunc(config, opts)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		return &gvkCaches{
+			gvkToCache:   gvkToCache,
+			defaultCache: defaultCache,
+			scheme:       opts.Scheme,
+		}, nil
+	}
+}
+
+type gvkCaches struct {
+	gvkToCache   map[schema.GroupVersionKind]Cache
+	defaultCache Cache
```

Comment:
It would contain any objects whose GVK is not explicitly in the GVK cache mapping. So the idea is that if you have a use case where a specific GVK needs a custom cache, then you can use this implementation to specify that cache for the GVK you care about and another (default) cache for all other GVKs your clients use.